### PR TITLE
Implement an external misk-crypto key source

### PR DESCRIPTION
**Pull request by yan**
This implements allowing for multiple sources for misk-crypto keys, and implements an S3-backed key source.

### CODE REVIEW

File: misk-crypto/src/test/kotlin/misk/crypto/CryptoModuleTest.kt [hunk 1]
```diff
@@ -91,30 +96,14 @@ class CryptoModuleTest {
...
+    val keysetHandle = KeysetHandle.generateNew(
+        HybridKeyTemplates.ECIES_P256_HKDF_HMAC_SHA256_AES128_GCM).publicKeysetHandle
     val injector = getInjector(listOf(Pair("test-hybrid", keysetHandle)))
     val hybridEncryptKeyManager = injector.getInstance(HybridEncryptKeyManager::class.java)
     assertThat(hybridEncryptKeyManager).isNotNull
     assertThat(hybridEncryptKeyManager["test-hybrid"]).isNotNull
   }
 
-  @Test
-  fun testImportUnencryptedHybridPublicKey() {
```
Comment (yoavamit) [inline misk-crypto/src/test/kotlin/misk/crypto/CryptoModuleTest.kt:102]: why was this deleted?

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 2]
```diff
@@ -232,21 +234,45 @@ internal class RealTransacter private constructor(
           session.postCommit()
           result
         } catch (e: Throwable) {
+          var rethrow = e
```
Comment (yoavamit) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:237]: unrelated?

Comment (yoavamit) [reply misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:237]: looks like there's a few more unrelated file changes here.
If it's not strictly related to loading keys from S3, can we put those changes in a different PR?

**Review by yoavamit — COMMENTED** [review]
Mostly nits, other than that, LGTM!

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt (hunk 2 above)
Comment (yan) [reply misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:237]: Hm this shouldn't be part of this PR, i assume this was an unintended rebase.

**Review by AlexSzlavik — APPROVED** [review]
Love the work here. Besides a bunch of format nits, the main comment is that this could probably have been split into multiple PRs. There is a refactor (maybe even 2) in here, plus a new feature.

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt [hunk 3]
```diff
@@ -33,77 +40,113 @@ class CryptoModule(
...
     HybridConfig.register()
     StreamingAeadConfig.register()
 
-    val keyNames = config.keys.map { it.key_name }
-    val duplicateNames = keyNames - keyNames.distinct().toList()
-    check(duplicateNames.isEmpty()) {
-      "Found duplicate keys: [$duplicateNames]"
+    var keyNames = listOf<KeyAlias>()
+
+    // Various key providers should exist in the same namespace
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt:58]: What I tried to say here is that multiple bound key providers all resolve keys in the same namespace.

File: misk-crypto/src/test/kotlin/misk/crypto/CryptoModuleTest.kt (hunk 1 above)
Comment (yan) [reply misk-crypto/src/test/kotlin/misk/crypto/CryptoModuleTest.kt]: has been re-added, it was accidental.

File: misk-aws/src/main/kotlin/misk/s3/RealS3Module.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import javax.inject.Singleton
+
+class RealS3Module: KAbstractModule() {
+  override fun configure() {
+    requireBinding<AWSCredentialsProvider>()
+    requireBinding<AwsRegion>()
+  }
+
+  @Provides @Singleton
+  fun provideS3(awsRegion: AwsRegion, awsCredentialsProvider: AWSCredentialsProvider): AmazonS3 {
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-aws/src/main/kotlin/misk/s3/RealS3Module.kt:18]: This is a global provider and it hasn't been used in misk yet. Should this be named? I modeled it after `RealDynamoDbModule`

[the remark this answers is not in the document]
Comment (AlexSzlavik) [reply misk-aws/src/main/kotlin/misk/s3/RealS3Module.kt:18]: I guess this is just providing the S3 client, which you'd only ever have one of. When I read it earlier I thought this is providing a bucket.

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt [hunk 5]
```diff
@@ -49,76 +53,112 @@ class CryptoModule(
...
-        }
-        KeyType.PGP_ENCRYPT -> {
-          bind<PgpEncrypter>()
-              .annotatedWith(Names.named(key.key_name))
-              .toProvider(PgpEncrypterProvider(key))
-              .`in`(Singleton::class.java)
-        }
+    // Include all configured remote keys
+    config.external_data_keys?.let { external_data_keys ->
+      requireBinding<AmazonS3>()
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt:78]: This crates an explicit dependency on an injected object. Normally, if it's not provided, this will fail at object use/injection site, where as in this case it will fail at the injector creation time.

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt [hunk 6]
```diff
@@ -49,76 +53,112 @@ class CryptoModule(
...
+
+      bind(object : TypeLiteral<Map<KeyAlias, KeyType>>() {})
+          .annotatedWith(Names.named("all key aliases"))
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/CryptoModule.kt:81]: I was thinking about this as well; will rename to `all_key_aliases` just to not raise any more brows :)

File: misk-crypto/src/main/kotlin/misk/crypto/CryptoTestModule.kt [hunk 7]
```diff
@@ -31,76 +34,93 @@ import misk.inject.KAbstractModule
  * Instead, it'll generate a random keyset handle for each named key.
  */
 class CryptoTestModule(
-  private val keyNames: List<Key>
+  private val config: CryptoConfig? = null
 ) : KAbstractModule() {
 
   override fun configure() {
 
+    bind<KmsClient>().toInstance(FakeKmsClient())
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/CryptoTestModule.kt:42]: It was, this just got hoisted to top of `configure()` instead of after all the registers.

File: misk-crypto/src/main/kotlin/misk/crypto/KeyProviders.kt [hunk 8]
```diff
@@ -26,17 +26,18 @@ import com.google.crypto.tink.streamingaead.StreamingAeadFactory
...
     val KEK_TEMPLATE: KeyTemplate = AeadKeyTemplates.AES256_GCM
   }
 
-  @Inject private lateinit var kmsClient: KmsClient
+  @Inject lateinit var kmsClient: KmsClient
 
-  val logger by lazy { getLogger<KeyReader>() }
+  @Inject lateinit var keySources: Set<ExternalKeyManager>
+
+  private val logger by lazy { getLogger<KeyReader>() }
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/KeyProviders.kt:40]: Good catch, this is no longer necessary. While I was working on this, the structure of the testing modules was slightly different, and this logger ended up being initialized before the LogCollector module was, so it wasn't able to collect logs. It should be fine now though, I'll revert.

File: misk-crypto/src/main/kotlin/misk/crypto/LocalConfigKeyProvider.kt [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+class LocalConfigKeyProvider(
+  private val keys: List<Key>,
+  private val kmsUri: String
+) : ExternalKeyManager {
+
+  override val allKeyAliases: Map<KeyAlias, KeyType> =
+      keys.map { key -> key.key_name to key.key_type }.toMap()
+
+  override fun getKeyByAlias(alias: KeyAlias): Key? {
+    return keys.find { key -> key.key_name == alias }?.copy(kms_uri = kmsUri)
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/LocalConfigKeyProvider.kt:18]: So this is to bridge between local and service keys and external keys. With local keys, we can have a single master key protect multiple service keys, and is generally passed alongside multiple keys. With external keys, each keyset gets its own, so this is a way to bridge between the two and standardize on a per-keyset instance.

File: misk-crypto/src/main/kotlin/misk/crypto/S3ExternalKeyManager.kt [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    allKeyAliases.mapValues { (alias, type) ->
+      logger.info { "registering external key: $alias" }
+      getRemoteKey(alias, type)
+    }.toMap(retrievedKeys)
+    retrievedKeys
+  }
+
+  override fun getKeyByAlias(alias: KeyAlias) = keys[alias]
+
+  override fun onKeyUpdated(cb: (KeyAlias, KeysetHandle) -> Unit) = false
```
[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/S3ExternalKeyManager.kt:90]: Not yet, but this is functionality that we might need once we start supporting key rotations. The interface is easy to ignore (just return false).

[the remark this answers is not in the document]
Comment (AlexSzlavik) [reply misk-crypto/src/main/kotlin/misk/crypto/S3ExternalKeyManager.kt:90]: Hrm, I don't think it's good design to add functionality that is unused. Generally we should implement things we need, not what we think we might need, my $0.02 to avoid clutter / bloated APIs

[the remark this answers is not in the document]
Comment (yan) [reply misk-crypto/src/main/kotlin/misk/crypto/S3ExternalKeyManager.kt:90]: I was debating taking it out, and this settles it. I'll remove it.
